arch/arm/core/cortex_m/mpu/Kconfig

```diff
@@ -38,3 +38,10 @@ config NXP_MPU
 	default n
 	help
 	  MCU has NXP MPU
+
+config MPU_ALLOW_FLASH_WRITE
+	bool "Add MPU access to write to flash"
+	depends on NXP_MPU
+	default n
```

Comment:
I think I hear what you're saying.

Of course, by enabling a driver, the user expects all of the routines it declares to work, but there is potentially some conflict between the flash and MPU drivers here, depending on what you mean by "level of security" (e.g., attacks on flash involving privileged code execution aren't stopped by MPUs, since they can disable the MPU before writing flash).

It seemed like resolving that in favor of security or least-surprise was the decision.

But @galak, I think you're saying:

- add a flash config option, FLASH_DISABLE_WRITE_SUPPORT, which is default y
- don't declare flash_write() or flash_erase() in flash.h unless FLASH_DISABLE_WRITE_SUPPORT is unset
- MPU disables flash writes if and only if FLASH_DISABLE_WRITE_SUPPORT=y

That will cause churn for existing flash users, but seems like it addresses all the concerns: users get everything they ask for, and MPU disables flash writes unless explicitly requested not to.

What do you think, @agross-linaro?